lib/usagereporter/teleport/types.go

```diff
@@ -407,26 +410,43 @@ func (u *ResourceHeartbeatEvent) Anonymize(a utils.Anonymizer) prehogv1.SubmitEv
 	}
 }
 
+type UserMetadataClient interface {
+	GetUsername(context.Context) (string, error)
+	GetUserIsSSO(context.Context, string) (bool, error)
+}
+
 // ConvertUsageEvent converts a usage event from an API object into an
 // anonymizable event. All events that can be submitted externally via the Auth
 // API need to be defined here.
-func ConvertUsageEvent(event *usageeventsv1.UsageEventOneOf, identityUsername string) (Anonymizable, error) {
+func ConvertUsageEvent(ctx context.Context, event *usageeventsv1.UsageEventOneOf, clt UserMetadataClient) (Anonymizable, error) {
 	// Note: events (especially pre-registration) that embed a username of their
 	// own should generally pass that through rather than using the identity
 	// username provided to the function. It may be the username of a Teleport
 	// component (e.g. proxy) rather than the end user.
 
 	switch e := event.GetEvent().(type) {
 	case *usageeventsv1.UsageEventOneOf_UiBannerClick:
+		identityUsername, err := clt.GetUsername(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
```

Comment:
This gets repeated a lot; would it be possible to do it once for all events?

Reply:
Before this PR changes it was doing it once for all events.
It is not required for all events, so we were doing this call even its value was not used.

Given that we need another param (IsSSO), I ended up removing the concrete value and we are now using an interface to resolve both variables.

I agree this gets repeated a lot, however the execution flow should be slightly faster.

I can add a switch case to call `GetUsername` and/or `GetUserIsSSO` before building the actual event
Then, the current switch case will only use the `identityUsername` and/or `isSSO` instead of calling the functions.

Something like this:
```go
func ConvertUsageEvent(ctx context.Context, event *usageeventsv1.UsageEventOneOf, clt UserMetadataClient) (Anonymizable, error) {
	var err error
	var identityUsername string
	var isSSOUser bool

	switch event.GetEvent().(type) {
	case *usageeventsv1.UsageEventOneOf_UiBannerClick,
		*usageeventsv1.UsageEventOneOf_UiOnboardAddFirstResourceClick,
		// all of the events that require IdentityUsername
		// ...
		*usageeventsv1.UsageEventOneOf_UiOnboardAddFirstResourceLaterClick:
		identityUsername, err = clt.GetUsername(ctx)
		if err != nil {
			return nil, trace.Wrap(err)
		}
	case *usageeventsv1.UsageEventOneOf_UiDiscoverStartedEvent,
		*usageeventsv1.UsageEventOneOf_UiDiscoverResourceSelectionEvent,
		// all of the events that require IdentityUsername and IsSSO
		// ...
		*usageeventsv1.UsageEventOneOf_UiDiscoverDeployServiceEvent:
		identityUsername, err = clt.GetUsername(ctx)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		isSSOUser, err = clt.GetUserIsSSO(ctx, identityUsername)
		if err != nil {
			return nil, trace.Wrap(err)
		}
	}

	switch e := event.GetEvent().(type) {
	case *usageeventsv1.UsageEventOneOf_UiBannerClick:
		return &UIBannerClickEvent{
			UserName: identityUsername,
			Alert:    e.UiBannerClick.Alert,
		}, nil
		//...
	
	case *usageeventsv1.UsageEventOneOf_UiDiscoverStartedEvent:
		ret := &UIDiscoverStartedEvent{
			Metadata: discoverMetadataToPrehog(e.UiDiscoverStartedEvent.Metadata, identityUsername, isSSOUser),
			Status:   discoverStatusToPrehog(e.UiDiscoverStartedEvent.Status),
		}
		if err := ret.CheckAndSetDefaults(); err != nil {
			return nil, trace.Wrap(err)
		}

		return ret, nil
		// ...
```

Being honest here, I'm not fond of either approach.
Maybe we should think a bit more about this.
What do you think? Do you have another suggestion?